### PR TITLE
Hotfix for 372: Use JSONEncoder in json.dumps

### PR DESCRIPTION
**Pull request by aranke**
Hotfix for https://app.incident.io/dbt-labs/incidents/372

### Problem

JSON serialization in dbt Cloud is failing on complex data types like `datetime`.

### Solution

Encode data with `dbt.utils.JSONEncoder`

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me  
- [x] I have run this code in development and it appears to resolve the stated issue  
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] This PR has no interface changes (e.g. macros, cli, logs, json artifacts, config files, adapter interface, etc) or this PR has already received feedback and approval from Product or DX

### CODE REVIEW

File: core/dbt/utils.py [hunk 1]
```diff
@@ -337,15 +337,16 @@ class JSONEncoder(json.JSONEncoder):
...
+        elif isinstance(obj, jinja2.Undefined):
             return ""
-        if hasattr(obj, "to_dict"):
+        elif hasattr(obj, "to_dict"):
             # if we have a to_dict we should try to serialize the result of
             # that!
             return obj.to_dict(omit_none=True)
-        return super().default(obj)
+        else:
+            return super().default(obj)
```
Comment (MichelleArk) [inline core/dbt/utils.py:349]: thank you for leaving the campground cleaner than you found it ✨⛺

**Review by MichelleArk — APPROVED** [review]
Looks good! Non blocking comment - could you please fill in the problem context in the PR description for posterity?

**Review by aaronraff — COMMENTED** [review]
Should this be using the forgiving encoder, or will the regular encoder be sufficient?
https://github.com/dbt-labs/dbt-core/blob/a1b067c683212bccd9bc10f6e52bcdd168ccf7ec/core/dbt/utils.py#L351

**Review by peter-bertuglia — APPROVED** [review]

**Review by aaronraff — APPROVED** [review]

**Review by gshank — COMMENTED** [review]
It looks like the JSONEncoder will handle decimal, datetime, jinja undefined and Exception types, but the ForgivingJSONEncoder will take any field that gets a TypeError and stringify it.  I haven't seen a list of the things which are blowing up in the json logs... do they all fall in that list of things handled by the JSONEncoder? For current code those things aren't going to make it through the logging event init code, but for older code wouldn't we want the Forgiving JSONEncoder? Seems like that was probably the reason that it was used in dbt-rpc.

**Comment by gshank** [comment]
If some other non-serializable class is passed to the old logging code, we'd end up with the same issue that this is supposed to address. I think it would be safer on older versions to used the ForgivingJSONEncoder. We'd rather not see Round 3 of json logging error whack-a-mole.

**Review by gshank — APPROVED** [review]

**Pull request merged**